Rework third chapter (#41)
0xAX/asm: content/asm_3.md

# Journey through the stack

In the [previous post](asm_2.md), we started to learn the basics of the x86_64 architecture. One of the most crucial concepts we learned was the [stack](https://en.wikipedia.org/wiki/Stack-based_memory_allocation). In this chapter, we will explore more examples of stack usage.

> Let's start with a quick remainder: the stack is a special memory region that operates on the LIFO (last-in, first-out) principle. In the x86_64 architecture, we have sixteen general-purpose registers for temporary data storage: `rax`, `rbx`, `rcx`, `rdx`, `rdi`, `rsi`, `rbp`, `rsp`, and from `r8` to `r15`. However, for some applications, this might not be enough. One way to overcome this limitation is by using the stack.

Reviewer: ```suggestion
Let's start with a quick reminder: the stack is a special memory region that operates on the LIFO (last-in, first-out) principle. In the x86_64 architecture, we have sixteen general-purpose registers for temporary data storage: `rax`, `rbx`, `rcx`, `rdx`, `rdi`, `rsi`, `rbp`, `rsp`, and from `r8` to `r15`. However, for some applications, this might not be enough. One way to overcome this limitation is by using the stack.
```

section .text

_start:
        ;; Put 1 to the rax register
		mov rax, 1
        ;; Call the incRax subroutine
		call incRax
        ;; Compare the value in the rax register with 2
		cmp rax, 2
        ;; Jump to the 'exit' label if not equal
		jne exit
		;;
		;; Otherwise, perform another action.
		;;

incRax:
        ;; Increment the value of the rax register
		inc rax

>         ;; Return from the incRax subroutine

Reviewer 2: indentation

- Number of bytes to subtract from the `rsp` register to allocate space on the stack.
- Number of stack frame levels in nested calls.

These instructions are considered "outdated" because of performance issues and the usual function prologue and epilogue are used, but still work because of backward compatibility.

The next familiar instruction that affects the stack is the `syscall` instruction. In some aspects, it is similar to the `call` instruction, with one key difference: the function to be called is located in kernel space. The return from a system call and the stack clean-up are executed using the `sysret` instruction.

In the previous post, we mentioned that there are other types of registers besides the general purpose registers. One such register is `rflags` where the CPU stores its current state. In the next posts, we will learn more about it. For now, we must know that the x86_64 processor provides the following two commands that affect the stack:

- `pushf` - pushes the `rflags` register into the stack.
- `popf` - pops the top value from the stack and stores it in the `rflags` register.

## Example

> After we went through some theory it is time to write some code! Let's see another one example that should make us more confident with the assembly programming. This time we will write a simple program, which will get [two command line arguments](https://en.wikipedia.org/wiki/Command-line_interface#Arguments), try to calculate sum of the given values and print the result.

Author: Yes, it is good to add :+1:

> [!NOTE]
> For the sake of simplification we will skip the check that numeric values are given in the command line arguments and do not do any checks for overflow. You may do it as your homework.

Before any explanation, first of all let's take a look at the whole code:

```assembly
;; Definition of the .data section
section .data
    ;; Number of `sys_write` system call
	SYS_WRITE equ 1
    ;; Number of `sys_exit` system call
	SYS_EXIT equ 60
    ;; Number of the standard output file descriptor
	STD_OUT	equ 1
    ;; Exit code from the program. The 0 status code is a success.
	EXIT_CODE equ 0
    ;; ASCII code of the new line symbol ('\n')
	NEW_LINE db 0xa
    ;; Error message that is printed in a case of not enough command line arguments
	WRONG_ARGC_MSG	db "Error: expected two command line argument", 0xa
    ;; Length of the WRONG_ARGC_MSG message

>     WRONG_ARGC_MSG_LEN equ 42

Reviewer 2: indentation

str_to_int:
	;; Set the value of the rax register to 0. It will store the result.
	xor rax, rax
	;; base for multiplication
	mov rcx,  10
__repeat:
	;; Compare the first element in the given string with the NUL terminator (end of string).
	cmp [rsi], byte 0
	;; If we reached the end of the string, return from the procedure. The result is stored in the rax register.
	je __return
	;; Move the current character from the command line argument to the bl register.
	mov bl, [rsi]
	;; Subtract the value 48 from the ASCII code of the current character.
    ;; This will give us the numeric value of the character.
	sub bl, 48
	;; Multiple our result number by 10 to get the place for the next digit.
	mul rcx
	;; Add the next digit to our result number.
	add rax, rbx
	;; Move to the next character in the command line argument string.
	inc rsi
	;; Repeat until we do not reach the end of the string.
	jmp __repeat
__return:
    ;; Return from the str_to_int procedure.
	ret

;; Convert the sum to string and print it on the screen.
int_to_str:
	;; High part of the dividend. The low part is in the rax register.
	mov rdx, 0
	;; Set the divisor to 10.
	mov rbx, 10

> 	;; Divide the sum (rax from rax) to 10. Reminder will be stored in the rdx register.

Author: Thank you for suggestion. Applied along with other comments. Thank you one more time.

In addition to preserving the return address, the stack is also used to access the function parameters and local variables. As you may recall from the previous chapter, the [System V AMD64 ABI](https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf) document specifies that the first six function parameters are passed in registers.

These registers are:

- `rdi` - used to pass the first argument to a function.
- `rsi` - used to pass the second argument to a function.
- `rdx` - used to pass the third argument to a function.
- `r10` - used to pass the fourth argument to a function.
- `r8` - used to pass the fifth argument to a function.
- `r9` - used to pass the sixth argument to a function.

Local variables are also accessed using the stack. For example, let's take a look at the following C function that doubles its parameter:

```C

> // The __ prefix in the __dobule function name to not mix it with the `double` data type.

Reviewer: ```suggestion
// The "__" prefix in the `__double` function name is used to avoid confusion with the `double` data type.
```

int __double(int a) {
    int two = 2;

    return a * two;
}
```

If we compile this function and take a look at the assembly output, we will see something like this:

```assembly
__double(int):
        ;; Preserve the base pointer
        push    rbp
        ;; Set the new frame base pointer
        mov     rbp, rsp

>         ;; Put the value of the first parameter of the function from the edi register on the stack with the location rbp - 20 bytes.

Reviewer: The GitHub preview cuts the sentence after the word "location":

<img width="914" alt="Screenshot 2025-01-12 at 10 50 42" src="https://github.com/user-attachments/assets/33e0bd78-7ddb-4696-8b67-a3da444d27c5" />

Not sure how critical it is, though 🤔 
We can see the full sentence after scrolling a bit to the right.

        ;; Put the 2 to on the stack with the location rbp - 4 bytes.
        mov     DWORD PTR [rbp-4], 2
        ;; Put the values of the first parameter of the function to the eax register.
        mov     eax, DWORD PTR [rbp-20]
        ;; Multiple the value of the eax register to 2 and store the result in the eax register.
        imul    eax, DWORD PTR [rbp-4]
        ;; Restore base pointer.
        pop     rbp
        ;; Exit from the __dobule function.
        ret
```

After the first two lines of the `__double` function the stack frame for this function is set and looks like:

![asm-3-stack-fram-of__double-1](./assets/asm-3-stack-of__double-1.svg)

The third instruction of the function `__double` puts the first parameter of this function to the stack with offset `-20`. After this we may see that the value `2` which is the value of the local variable `two` is also put onto the stack with the offset `-4`. The stack frame of our function for this moment should look like this: 

![asm-3-stack-fram-of__double-2](./assets/asm-3-stack-of__double-2.svg)

> After this we put the value from the stack with the offset `-20` (the value of the function's parameter) to the register eax and multiply it by `2` which is located on the stack with the offset `-4`. The result of the multiplication will be in the register eax. This simple example shows how stack is used to access and parameters and local variables of the function.

Reviewer: ```suggestion
Finally, we put the value from the stack at offset `-20` (the value of the function's parameter) into the `eax` register and multiply it by `2`, which is located on the stack at offset `-4`. The result of the multiplication is then stored in the `eax` register. This simple example shows how the stack is used to access both parameters and local variables of a function.
```

Finally, we put the value from the stack at offset `-20` (the value of the function's parameter) into the `eax` register and multiply it by `2`, which is located on the stack at offset `-4`. The result of the multiplication is then stored in the `eax` register. This simple example shows how the stack is used to access both parameters and local variables of a function.

## Stack operations

We already have seen two assembly instructions that affects the current state of the stack:

- `push` - pushes the operand into the stack.
- `pop` - pops the top value from the stack.

> x86_64 processors provide additional instruction that brings affect on the stack. Besides those instruction we also have seen familiar to us:

Reviewer: ```suggestion
x86_64 processors provide additional instructions that affect the stack. In addition to these, we’ve also seen instructions that are already familiar to us:
```

> - `call`
> - `ret`

Reviewer: ```suggestion
- `call` - calls the given procedure. It affects the stack by saving the return address before the call.
- `ret` - exits the given procedure. It affects the stack by removing the return address and transferring the execution flow back to it. 
```

> The first one instruction calls the given procedure. It affects stack by saving the return address on the stack before call. The second instruction is an "exit" from the given procedure. It affects the stack by removing the return address from the stack and transferring the execution flow to it. 

Reviewer: ```suggestion

```

I would explain these instructions right in the list as in the comment above. If you like this idea, please remove this paragraph.

Author: Will remove

> In the [previous post](asm_2.md) we got familiar with the with such a concepts as [function prologue and epilogue](https://en.wikipedia.org/wiki/Function_prologue_and_epilogue). These are special instructions that we usually can meet in the beginning and in the end of the function:

Reviewer: ```suggestion
In the [previous post](asm_2.md), we became familiar with concepts such as the [function prologue and epilogue](https://en.wikipedia.org/wiki/Function_prologue_and_epilogue). These are special instructions typically found at the beginning and end of a function:
```

```assembly
foo:
        ;; Function prologue
        push %rbp
        mov     rbp, rsp

        ;;
        ;; Function body
        ;;

        ;; Function epilogue
        mov rsp, rbp
        pop
```

> These two could be replaced with special instructions: `enter N, 0` and `leave`. The first instruction has two operands: 

Reviewer: ```suggestion
These two can be replaced with special instructions: `enter N, 0` and `leave`. The `enter` instruction has two operands: 
```

> - Number of bytes that needs to be subtracted from the `rsp` register to allocate space on stack.
> - Number of levels of stack frames in nested calls.

Reviewer: ```suggestion
- Number of bytes to subtract from the `rsp` register to allocate space on the stack.
- Number of stack frame levels in nested calls.
```

> These both instructions are considered "outdated" but still will work because of backward compatibility.

Reviewer: ```suggestion
These instructions are considered "outdated" but still work because of backward compatibility.
```

Reviewer: Why are they outdated? By what are they replaced?

Author: Good note, added the reason

> The next already familiar to us instruction that affects the stack is the `syscall` instruction. In some aspects it is similar to the `call` instruction with the one of the most significant difference is that the function that is going to be called is located in the kernel space. The return from a system call and the stack clean-up is executed with the help of the `sysret` instruction.

Reviewer: ```suggestion
The next familiar instruction that affects the stack is the `syscall` instruction. In some aspects, it is similar to the `call` instruction, with one key difference: the function to be called is located in kernel space. The return from a system call and the stack clean-up are executed using the `sysret` instruction.
```

> In the previous post, we mentioned that besides the general purpose registers, the other types of registers exists. One of such type of registers is `rflags`. In basic words it is a register where CPU stores its current state. In the next posts we will know more details about this type of register but for now we must know that an x86_64 process provide the two following command that affect the stack:

Reviewer: ```suggestion
In the previous post, we mentioned that there are other types of registers besides the general purpose registers. One such register is `rflags` where the CPU stores its current state. In the next posts, we will learn more about it. For now, we must know that the x86_64 processor provides the following two commands that affect the stack:
```

> - `pushf` - pushes the `rflags` register into the stack.
> - `popf` - pops the top value from the stack and stores the value in the `rflags` register.

Reviewer: ```suggestion
- `pushf` - pushes the `rflags` register into the stack.
- `popf` - pops the top value from the stack and stores it in the `rflags` register.
```

2. Convert the result back to string and print it on the screen.

In the next two sections, we will see a detailed explanation of these steps.

### Converting a string to an integer

As the command-line arguments of each program are represented as strings, first we need to convert our command-line arguments to numbers to calculate their sum. To convert a given string to a number, we will use a simple algorithm:

1. Create an accumulator to store an intermediate result while converting the string into its numeric representation.
2. Take the first byte of the string and subtract the value `48` from it. Each byte in a string is an [ASCII](https://en.wikipedia.org/wiki/ASCII) symbol with its own code. The symbol 0 has code `48`, the symbol 1 has code `49`, and so on. If we subtract `48` from the ASCII code of the given symbol, we get an integer representation of the current digit from the given string.
3. As soon as we know the current digit, we multiply our accumulator from step 1 by 10 and add to it the digit that we got in step 2.
4. Move to the next symbol in the given string and repeat steps 2 and 3 if it is not the end of the string (`\0` symbol).

> Returning to the table from the section above, we may see that pointers to the command line arguments are located on the stack right above the number of command line arguments. So if we fetch the first value from the stack after we already fetched the number of arguments, it will be a pointer to the string which is the first command line argument. If we will pop the next value from the stack, it will be the second command line argument passed to the program.

Reviewer: ```suggestion
Returning to the table from the section above, we can see that pointers to the command-line arguments are located on the stack right above the number of command-line arguments. So, after we pop the number of arguments (ARGC), the stack pointer will point to the address of the first command-line argument (ARGV[0]). If we pop the next value from the stack, it will point to the second command-line argument (ARGV[1]) passed to the program.
```

> Now if we will take a look at the `str_to_int` procedure it should be clear without any additional details:

Reviewer: ```suggestion
Now the `str_to_int` procedure should be more clear:
```

	pop rsi
	;; Convert the second command line argument to an integer number.
	call str_to_int
	;; Store the result in the r11 register.
	mov r11, rax

	...
	...
	...

;; Convert the command line argument to the integer number.
str_to_int:
	;; Set the value of the rax register to 0. It will store the result.
	xor rax, rax
	;; base for multiplication
	mov rcx,  10
__repeat:
	;; Compare the first element in the given string with the NUL terminator (end of string).
	cmp [rsi], byte 0
	;; If we reached the end of the string, return from the procedure. The result is stored in the rax register.
	je __return
	;; Move the current character from the command line argument to the bl register.
	mov bl, [rsi]
	;; Subtract the value 48 from the ASCII code of the current character.
	;; This will give us the numeric value of the character.
	sub bl, 48
	;; Multiple our result number by 10 to get the place for the next digit.
	mul rcx
	;; Add the next digit to our result number.
	add rax, rbx
	;; Move to the next character in the command line argument string.
	inc rsi
	;; Repeat until we do not reach the end of the string.
	jmp __repeat
__return:
	;; Return from the str_to_int procedure.
	ret
```

> As soon as we converted both command line arguments to integer numbers, we can calculate their sum:

Reviewer: ```suggestion
As soon as we converted both command-line arguments to integer numbers, we can calculate their sum:
```

```assembly
	;; Calculate the sum of the arguments. The result will be stored in the r10 register.
	add r10, r11
```

> Since we have our result, we just need to print it. But before printing it we have to convert the numeric result to string. This we will see in the next section.

Reviewer: ```suggestion
Now that we have our result, we just need to print it. But before printing it, we have to convert the numeric result back to a string. This we will learn in the next section.
```

> ### Converting integer to string

Reviewer: ```suggestion
### Converting an integer to a string
```

	;; The div instruction works as div operand => rdx:rax / operand.
	;; The remainder is stored in rdx and the quotient in rax.
	mov rdx, 0
	;; Set the divisor to 10.
	mov rbx, 10
	;; Divide the sum stored in `rax`, resulting quotient will be stored in `rax`,
	;; and the remainder will be stored in `rdx` register.
	div rbx
	;; Add 48 to the remainder to get a string ASCII representation of the number value.
	add rdx, 48
	;; Store the remainder on the stack.
	push rdx
	;; Increase the counter.
	inc rcx
	;; Compare the rest of the sum with zero.
	cmp rax, 0x0
	;; If it is not zero yet, continue to convert it to string.
	jne int_to_str
	;; Otherwise print the result.
	jmp printResult
```

Before jumping to the `int_to_str` sobroutine, we need to do some preparations. As you may see we put the value of our sum in the `rax` register and initialize the counter (`rcx` register) with zero. This counter will store the number of symbols in the our future string. Note that we are using new instruction to initialize the counter - `xor`. This instruction is a [bitwise XOR](https://en.wikipedia.org/wiki/Bitwise_operation#XOR) operator which resets bits of the operands to 0 if they are the same.

> The algorithm of the `int_to_str` sobroutine is pretty simple as well. We divide our number by `10` to get the digit and add the value `48` to the result of the division. Remember about ASCII codes? If yes it should be clear why we are doing it. As soon as we got the symbolic representation of the current digit we push it on the stack. As soon as the given digit is converted we increase our counter of numbers of symbols within the string and check our sum number. If it is zero it means we have the resulted string. If not, we just repeat the all operations.

Reviewer: If yes, it should be clear why we are doing it:
1. As soon as we get the symbolic representation of the current digit, we push it on the stack. 
2. When the given digit is converted, we increase our counter that represents the number of characters within the string. 
3. After that, we check the sum number. If it is zero, we have the resulting string. If not, we repeat all operations.

> The algorithm of the `int_to_str` sobroutine is pretty simple as well. We divide our number by `10` to get the digit and add the value `48` to the result of the division. Remember about ASCII codes? If yes it should be clear why we are doing it. As soon as we got the symbolic representation of the current digit we push it on the stack. As soon as the given digit is converted we increase our counter of numbers of symbols within the string and check our sum number. If it is zero it means we have the resulted string. If not, we just repeat the all operations.

Reviewer: ^ consider the alternative formating of text in the form of ordered list

	mov rcx, 8
	;; Calculate the number of bytes in the given string by multiplying rax by 8.
	;; The result will be stored in the rax register.
	mul rcx

	;; Set the third argument to the length of the result string to print.
	mov rdx, rax
	;; Specify the system call number (1 is `sys_write`).
	mov rax, SYS_WRITE
	;; Set the first argument of `sys_write` to 1 (`stdout`).
	mov rdi, STD_OUT
	;; Set the second argument of `sys_write` to the reference of the result string to print.
	mov rsi, rsp
	;; Call the `sys_write` system call.
	syscall

	;; Specify the system call number (1 is `sys_write`).
	mov rax, SYS_WRITE
	;; Set the first argument of `sys_write` to 1 (`stdout`).
	mov rdi, STD_OUT
	;; Set the second argument of `sys_write` to the reference of the `NWE_LINE` variable.
	mov rsi, NEW_LINE
	;; Set the third argument to the length of the `NEW_LINE` variable's value (1 byte).
	mov rdx, 1
	;; Call the `sys_write` system call.
	syscall

exit:
	;; Specify the number of the system call (60 is `sys_exit`).
	mov rax, SYS_EXIT
	;; Set the first argument of `sys_exit` to 0. The 0 status code is a success.
	mov rdi, EXIT_CODE
	;; Call the `sys_exit` system call.
	syscall
```

> Most of this code should be already well understandable for you as the most significant part of it consists of the initialization of data for the call of the `sys_write` and `sys_exit` exit calls. The most interesting part should be first four lines of code of the `printResult` subroutine. As you may remember the one of the parameters of the `sys_write` system call is a length of the string that we want to print on the screen. We have this number as we maintained a counter of symbols during converting the numeric sum to the string. This counter was stored in the `rcx` register. Our string is located on the stack. We pushed each digit with the `push` operator. But the `push` operator pushes `64` bits (or `8` bytes) while our symbol is only 1 byte. To get the whole length of the string for printing, we should multiple the number of symbols to `8`. This will give us the length of the string that we can use as a third argument of the `sys_write` system call.

Reviewer: ```suggestion
Most of this code should already be understandable, as it mainly consists of the data initialization for the `sys_write` and `sys_exit` system calls. The most interesting part is the first four lines of the `printResult` subroutine. As you may remember, one of the parameters of the `sys_write` system call is the length of the string we want to print on the screen. We have this number because we maintained a counter of characters while converting the numeric sum to a string. This counter was stored in the `rcx` register. Our string is located on the stack, where we pushed each digit using the `push` operator. However, the `push` operator pushes `64` bits (or `8` bytes), while our symbol is only 1 byte. To calculate the total length of the string for printing, we should multiply the number of symbols by `8`. This will give us the length of the string that we can use as the third argument of the `sys_write` system call.
```

> As soon as all parameters of the system calls are ready, we can pass them as arguments to print the sum and print new line after it.

Reviewer: ```suggestion
Once all parameters of both system calls are ready, we can pass them as arguments to print the sum followed by a new line.
```

> Let's build our program with the usual commands:

Reviewer: ```suggestion
Now, let's build our program with the usual commands:
```

```bash
$ nasm -f elf64 -o stack.o stack.asm
$ ld -o stack stack.o
```

> And try to run it:

Reviewer: ```suggestion
Then, try to run it:
```

}

int main() {
    foo();
    printf("Program exited successfully\n");
    return 0;
}
```

If we build and run this program, we'll see the following error instead of the `Program exited successfully` string:

```bash
$ ./test
Enter text: 123456789
*** stack smashing detected ***: terminated
Aborted (core dumped)
```

The reason for this error is that we put on the stack a value bigger than our 8-byte buffer. Happily, instead of overwriting the return address or segmentation fault error, we get a `stack smashing detected` error. This check is done by a modern compiler to prevent overwriting critical data. There are also other techniques in modern compilers and operating system kernels to mitigate vulnerabilities related to stack, like:

- [Stack canaries](https://en.wikipedia.org/wiki/Buffer_overflow_protection#Canaries)
- [ASLR](https://en.wikipedia.org/wiki/Address_space_layout_randomization)
- [Non-executable stack](https://en.wikipedia.org/wiki/Executable-space_protection)
- And others...

Despite all of these techniques may help you to protect your programs from stack-related errors, you should be careful, especially with the external data that your program receives.

> This example might be a little bit artificial as unlikely you are going to use the `gets` function in your code. The [manual page](https://man7.org/linux/man-pages/man3/gets.3.html) of this function says:

Reviewer: ```suggestion
The C function example above might be a little bit artificial as unlikely you are going to use the `gets` function in your code. The [manual page](https://man7.org/linux/man-pages/man3/gets.3.html) of this function says:
```

	call str_to_int
	;; Store the result in the r11 register.
	mov r11, rax

	;; Calculate the sum of the arguments. The result will be stored in the r10 register.
	add r10, r11
	;; Move the sum value to the rax register.
	mov rax, r10
	;; Initialize counter by resetting it to 0. It will store the length of the result string.
	xor rcx, rcx
	;; Convert the sum from a number to a string to print the result to the standard output.
	jmp int_to_str

;; Print the error message if not enough command line arguments.
argcError:
	;; Specify the system call number (1 is `sys_write`).
	mov rax, SYS_WRITE
	;; Set the first argument of `sys_write` to 1 (`stdout`).
	mov rdi, STD_OUT
	;; Set the second argument of `sys_write` to the reference of the `WRONG_ARGC_MSG` variable.
	mov rsi, WRONG_ARGC_MSG
	;; Set the third argument to the length of the `WRONG_ARGC_MSG` variable's value.
	mov rdx, WRONG_ARGC_MSG_LEN
	;; Call the `sys_write` system call.
	syscall
	;; Go to the exit of the program.
	jmp exit

;; Convert the command line argument to the integer number.
str_to_int:
	;; Set the value of the rax register to 0. It will store the result.
	xor rax, rax

> 	;; base for multiplication

Reviewer: ```suggestion
	;; Base for multiplication
```

Returning to the table from the section above, we can see that pointers to the command-line arguments are located on the stack right above the number of command-line arguments. So, after we pop the number of arguments (ARGC), the stack pointer will point to the address of the first command-line argument (ARGV[0]). If we pop the next value from the stack, it will point to the second command-line argument (ARGV[1]) passed to the program.

Now the `str_to_int` procedure should be more clear:

```assembly
	;; Fetch the first command line argument from the stack and store it in the rsi register.
	pop rsi
	;; Convert the first command line argument to an integer number.
	call str_to_int
	;; Store the result in the r10 register.
	mov r10, rax

	;; Fetch the second command line argument from the stack and store it in the rsi register.
	pop rsi
	;; Convert the second command line argument to an integer number.
	call str_to_int
	;; Store the result in the r11 register.
	mov r11, rax

	...
	...
	...

;; Convert the command line argument to the integer number.
str_to_int:
	;; Set the value of the rax register to 0. It will store the result.
	xor rax, rax

> 	;; base for multiplication

Reviewer: ```suggestion
	;; Base for multiplication
```

Once all parameters of both system calls are ready, we can pass them as arguments to print the sum followed by a new line.

Now, let's build our program with the usual commands:

```bash
$ nasm -f elf64 -o stack.o stack.asm
$ ld -o stack stack.o
```

Then, try to run it:

```bash
$  ./stack
Error: expected two command line argument
$ ./stack 5
Error: expected two command line argument
$ ./stack 5 10
15
```

Works as expected 🎉🎉🎉

## Security considerations

> As seen in this and the previous posts, the stack is a crucial concept used to manage function calls in our programs. Understanding how the stack memory is managed is important for writing programs with reusable functions and crucial for writing secure programs. The stack is a common source of security vulnerabilities, especially in low-level code and assembly routines. When you use `call` and `ret` instructions, the processor doesn’t verify if the return address is valid, but it just simply pops the address and jumps on it. One of the most common problems is the [stack overflow](https://en.wikipedia.org/wiki/Stack_overflow).

Reviewer: ```suggestion
As seen in this and the previous posts, the stack is a crucial concept used to manage function calls in our programs. Understanding how the stack memory is managed is important for writing programs with reusable functions and crucial for writing secure programs. The stack is a common source of security vulnerabilities, especially in low-level code and assembly routines. When you use `call` and `ret` instructions, the processor doesn’t verify if the return address is valid, but it simply pops the address and jumps on it. One of the most common problems is the [stack overflow](https://en.wikipedia.org/wiki/Stack_overflow).
```

Let's take a look at the simple C function (the function is written on C for simplicity):

```C
#include <stdio.h>
#include <string.h>

void foo() {
    char buffer[8];

    printf("Enter text: ");

    gets(buffer);
}

int main() {
    foo();
    printf("Program exited successfully\n");
    return 0;
}
```

> If we will try to build this program and run it, we'll see the following error instead of the *Program exited successfully* string:

Reviewer: ```suggestion
If we build and run this program, we'll see the following error instead of the `Program exited successfully` string:
```

```bash
$ ./test
Enter text: 123456789
*** stack smashing detected ***: terminated
Aborted (core dumped)
```

> The reason for this is that we put on the stack the value which is bigger than our 8 bytes buffer. Happily instead of overwriting of return address or segmentation fault error we have got "stack smashing detected" error. This check is done by the modern compiler to prevent overwriting of critical data. There are other techniques in modern compilers and operating system kernels to mitigate vulnerabilities related to stack, like:

Reviewer: ```suggestion
The reason for this error is that we put on the stack a value bigger than our 8-byte buffer. Happily, instead of overwriting the return address or segmentation fault error, we get a `stack smashing detected` error. This check is done by a modern compiler to prevent overwriting critical data. There are also other techniques in modern compilers and operating system kernels to mitigate vulnerabilities related to stack, like:
```

- [Stack canaries](https://en.wikipedia.org/wiki/Buffer_overflow_protection#Canaries)
- [ASLR](https://en.wikipedia.org/wiki/Address_space_layout_randomization)
- [Non-executable stack](https://en.wikipedia.org/wiki/Executable-space_protection)
- And others...

> In any cases, despite all of these techniques may help you to protect your programs from stack related errors, you should be careful, especially with the data that your program receives from outside.

Reviewer: ```suggestion
Despite all of these techniques may help you to protect your programs from stack-related errors, you should be careful, especially with the external data that your program receives.
```

> This example might be a little bit artificial as unlikely you are going to use the `gets` function in your code. The [manual page](https://man7.org/linux/man-pages/man3/gets.3.html) of this function says:

Reviewer: Which example? I feel like we lost a paragraph somewhere 🤔

> Never use gets().  Because it is impossible to tell without knowing the data in advance how many characters gets() will read, and because gets() will continue to store characters past the end of the buffer, it is extremely dangerous to use.  It has been used to break computer security.  Use fgets() instead.

> This example might seem a bit artificial as unlikely you are going to use the deprecated `gets` function. However, even with such an unrealistic example, real risks still exist — even if you avoid deprecated functions and use all the compiler’s safety features to protect your program.

Reviewer: And here we repeat the paragraph above 🤔

> This example might be a little bit artificial as unlikely you are going to use the `gets` function in your code. The [manual page](https://man7.org/linux/man-pages/man3/gets.3.html) of this function says:
> 
> > Never use gets().  Because it is impossible to tell without knowing the data in advance how many characters gets() will read, and because gets() will continue to store characters past the end of the buffer, it is extremely dangerous to use.  It has been used to break computer security.  Use fgets() instead.
> 
> Moreover this function is deprecated. But despite such artificial example, there can be real danger even if you are not using deprecated functions and use all the options of the compiler that help you to protect your program. 

Reviewer: ```suggestion
This example might seem a bit artificial as unlikely you are going to use the deprecated `gets` function. However, even with such an unrealistic example, real risks still exist — even if you avoid deprecated functions and use all the compiler’s safety features to protect your program.
```

Reviewer: too much "screen time" for `gets`

> The real world case when a wrong memory management led to serious consequences is [CVE-2017-1000253](https://nvd.nist.gov/vuln/detail/CVE-2017-1000253). This vulnerability was found in the Linux kernel and led to the [privilege escalation](https://en.wikipedia.org/wiki/Privilege_escalation). When the kernel runs a process, it needs to perform many different operations. One of such operations are to load the program into memory and initialize the stack. The binary itself is loaded and located below the stack memory. Besides that, there is a gap which is 128 megabytes. The loading of a big enough binary led to the situation when certain segments of the binary were loaded and mapped to this gap. The binaries with the big enough data segements (bingger than 128 megabytes) can end up mapped over the stack memory. All of this may lead to the situation when the [ELF .dynamic section](https://en.wikipedia.org/wiki/Executable_and_Linkable_Format) is overwritten by the stack. One may put a path to a shared library that will be loaded by the kernel and the code executed with the higher privileges. If you are interested in more details you can read the [report](https://www.qualys.com/2017/09/26/linux-pie-cve-2017-1000253/cve-2017-1000253.txt) and the [fix](https://github.com/torvalds/linux/commit/a87938b2e246b81b4fb713edb371a9fa3c5c3c86).

Reviewer: ```suggestion
The real-world case when wrong memory management led to serious consequences is [CVE-2017-1000253](https://nvd.nist.gov/vuln/detail/CVE-2017-1000253). This vulnerability was found in the Linux kernel and led to the [privilege escalation](https://en.wikipedia.org/wiki/Privilege_escalation). When the kernel runs a process, it needs to perform many different operations, such as loading the program into memory and initializing the stack. After the program is loaded and stack initialized, the program is located below the stack memory, with a 128-megabyte gap between them. However, when a large program is loaded, it can overwrite the stack memory. Under certain conditions, it may lead to privilege escalation. If you are interested in more details, you can read the [report](https://www.qualys.com/2017/09/26/linux-pie-cve-2017-1000253/cve-2017-1000253.txt) and the [fix](https://github.com/torvalds/linux/commit/a87938b2e246b81b4fb713edb371a9fa3c5c3c86).
```

> As we've seen, subtle bugs in stack layout can lead to serious vulnerabilities.

Reviewer: ```suggestion
As you can see, subtle bugs in stack layout can lead to serious vulnerabilities.
```

## Conclusion

> We’ve just written our third program using assembly — great job 🎉 In the next post, we’ll continue exploring assembly programming and see more details how to work with strings. If you have any questions or thoughts, feel free to reach out. See you in the next post!

Reviewer: ```suggestion
We’ve just written our third program using assembly — great job 🎉 In the next post, we’ll continue exploring assembly programming and see more details on how to work with strings. If you have any questions or thoughts, feel free to reach out. See you in the next post!
```
